File: flow/src/main/java/com/sequenceiq/cloudbreak/core/flow2/FlowLogService.java
package com.sequenceiq.cloudbreak.core.flow2;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Queue;
import java.util.Set;

import com.sequenceiq.cloudbreak.cloud.event.Payload;
import com.sequenceiq.cloudbreak.cloud.event.Selectable;
import com.sequenceiq.cloudbreak.domain.FlowChainLog;
import com.sequenceiq.cloudbreak.domain.FlowLog;
import com.sequenceiq.cloudbreak.service.TransactionService;

public interface FlowLogService {
    FlowLog save(String flowId, String flowChanId, String key, Payload payload, Map<Object, Object> variables, Class<?> flowType,
            FlowState currentState);

    Iterable<FlowLog> saveAll(Iterable<FlowLog> entities);

    FlowLog close(Long stackId, String flowId) throws TransactionService.TransactionExecutionException;

    FlowLog cancel(Long stackId, String flowId) throws TransactionService.TransactionExecutionException;

    FlowLog terminate(Long stackId, String flowId) throws TransactionService.TransactionExecutionException;

    void purgeTerminatedStacksFlowLogs() throws TransactionService.TransactionExecutionException;

    void saveChain(String flowChainId, String parentFlowChainId, Queue<Selectable> chain);

    void updateLastFlowLogStatus(FlowLog lastFlowLog, boolean failureEvent);

    boolean isOtherFlowRunning(Long stackId);

    boolean repeatedFlowState(FlowLog lastFlowLog, String event);

    void updateLastFlowLogPayload(FlowLog lastFlowLog, Payload payload, Map<Object, Object> variables);

    Optional<FlowLog> getLastFlowLog(String flowId);

    Set<String> findAllRunningNonTerminationFlowIdsByStackId(Long stackId);

    Optional<FlowLog> findFirstByFlowIdOrderByCreatedDesc(String flowId);

    Optional<FlowChainLog> findFirstByFlowChainIdOrderByCreatedDesc(String flowChainId);

    List<Object[]> findAllPending();

    Set<FlowLog> findAllUnassigned();

    Set<FlowLog> findAllByCloudbreakNodeId(String cloudbreakNodeId);

    List<FlowLog> findAllByStackIdOrderByCreatedDesc(Long id);

    Set<Long> findTerminatingStacksByCloudbreakNodeId(String id);

}

File: core/src/main/java/com/sequenceiq/cloudbreak/service/flowlog/FlowLogService.kt
package com.sequenceiq.cloudbreak.service.flowlog

import javax.inject.Inject
import javax.transaction.Transactional

import org.springframework.stereotype.Service

import com.sequenceiq.cloudbreak.cloud.event.Payload
import com.sequenceiq.cloudbreak.core.flow2.FlowState
import com.sequenceiq.cloudbreak.domain.FlowLog
import com.sequenceiq.cloudbreak.repository.FlowLogRepository

@Service
@Transactional
class FlowLogService {

    @Inject
    private val flowLogRepository: FlowLogRepository? = null

    fun save(flowId: String, key: String, payload: Payload, flowType: Class<*>, currentState: FlowState): FlowLog {
        val flowLog = FlowLog(payload.stackId, flowId, key, null, payload.javaClass, flowType, currentState.toString())
        return flowLogRepository!!.save(flowLog)
    }

    fun close(stackId: Long?, flowId: String): FlowLog {
        return finalize(stackId, flowId, "FINISHED")
    }

    fun cancel(stackId: Long?, flowId: String): FlowLog {
        return finalize(stackId, flowId, "CANCELLED")
    }

    fun terminate(stackId: Long?, flowId: String): FlowLog {
        return finalize(stackId, flowId, "TERMINATED")
    }

    private fun finalize(stackId: Long?, flowId: String, state: String): FlowLog {
        flowLogRepository!!.finalizeByFlowId(flowId)
        val flowLog = FlowLog(stackId, flowId, state, java.lang.Boolean.TRUE)
        return flowLogRepository.save(flowLog)
    }

    fun findAllRunningNonTerminationFlowIdsByStackId(stackId: Long?): Set<String> {
        return flowLogRepository!!.findAllRunningNonTerminationFlowIdsByStackId(stackId)
    }
}
